experimental/unitsconversion/src/converter.rs

```diff
@@ -0,0 +1,133 @@
+
+use crate::{
+    measureunit::{MeasureUnit, MeasureUnitParser},
+    provider::{MeasureUnitItem, UnitsInfoV1},
+    ConversionError,
+};
+use litemap::LiteMap;
+use zerovec::ZeroSlice;
+
+/// Represents the possible cases for the convertibility between two units.
+pub enum Convertibility {
+    /// The units are convertible.
+    /// For example, `meter` and `foot` are convertible.
+    Convertible,
+
+    /// The units are reciprocal.
+    /// For example, `meter-per-second` and `second-per-meter` are reciprocal.
+    Reciprocal,
+
+    /// The units are not convertible.
+    /// For example, `meter` and `second` are not convertible.
+    NotConvertible,
+}
+
+/// A factory for creating a converter.
+/// Also, it can check the convertibility between two units.
+pub struct ConverterFactory<'data> {
+    /// Contains the necessary data for the conversion factory.
+    payload: &'data UnitsInfoV1<'data>,
+}
+
+impl<'data> ConverterFactory<'data> {
+    #[cfg(feature = "datagen")]
+    pub fn from_payload(payload: &'data UnitsInfoV1<'data>) -> Self {
+        Self { payload }
+    }
+
+    pub fn parser(&self) -> MeasureUnitParser<'_> {
+        MeasureUnitParser::from_payload(&self.payload.units_conversion_trie)
+    }
+
+    /// Extract the convertibility from the given units in the form of CLDR identifiers.
+    pub fn extract_convertibility(
+        &self,
+        unit1: &MeasureUnit,
+        unit2: &MeasureUnit,
+    ) -> Result<Convertibility, ConversionError> {
+        let unit1 = &unit1.contained_units;
+        let unit2 = &unit2.contained_units;
+
+        struct DetermineConvertibility {
+            convertible: i16,
+            reciprocal: i16,
+        }
+
+        /// Inserting the non basic units into the map.
+        /// Thus means that from the given unis
+        fn insert_non_basic_units(
+            factory: &ConverterFactory,
+            units: &[MeasureUnitItem],
+            sign: i8,
+            map: &mut LiteMap<u16, DetermineConvertibility>,
+        ) -> Result<(), ConversionError> {
+            for item in units {
+                let items_from_item = factory
+                    .payload
+                    .convert_infos
+                    .get(item.unit_id as usize)
+                    .ok_or(ConversionError::InternalError)?;
+
+                insert_units_powers(items_from_item.basic_units(), item.power, sign, map)?;
+            }
+
+            Ok(())
+        }
+
+        fn insert_units_powers(
+            basic_units: &ZeroSlice<MeasureUnitItem>,
+            original_power: i8,
+            sign: i8,
+            map: &mut LiteMap<u16, DetermineConvertibility>,
+        ) -> Result<(), ConversionError> {
+            for item in basic_units.iter() {
+                let item_power = item
+                    .power
+                    .checked_mul(original_power)
+                    .ok_or(ConversionError::InternalError)?;
```

Comment:
This error case is if an `i8 x i8` multiplication overflows? Why not handle it by working with `i16`?

Reply:
good idea.